File: sdk/csharp/src/AnyVali/Schemas/BoolSchema.cs
namespace AnyVali.Schemas;

public sealed class BoolSchema : Schema
{
    internal override string GetCoercionTarget() => "bool";

    public new BoolSchema Default(object? value) => (BoolSchema)base.Default(value);
    public new BoolSchema Coerce(Parse.CoercionConfig? config = null) => (BoolSchema)base.Coerce(config);

    internal override object? Validate(object? input, ValidationContext ctx)
    {
        if (input is not bool)
        {
            ctx.Issues.Add(new ValidationIssue
            {
                Code = IssueCodes.InvalidType,
                Message = $"Expected boolean, received {DescribeType(input)}",
                Path = ctx.ClonePath(),
                Expected = "bool",
                Received = DescribeType(input),
            });
            return null;
        }
        return input;
    }

    internal override Dictionary<string, object?> ToNode()
    {
        var node = new Dictionary<string, object?> { ["kind"] = "bool" };
        AddDefaultAndCoercion(node);
        return node;
    }

    internal override Schema Clone() => new BoolSchema
    {
        DefaultValue = DefaultValue, CoercionCfg = CoercionCfg, IsPortable = IsPortable,
    };
}

File: sdk/ruby/lib/anyvali/schemas/bool_schema.rb
# frozen_string_literal: true

module AnyVali
  class BoolSchema < Schema
    def initialize(**kwargs)
      super(kind: "bool", **kwargs)
    end

    protected

    def validate(value, path, issues, context)
      unless value.is_a?(TrueClass) || value.is_a?(FalseClass)
        issues << ValidationIssue.new(
          code: IssueCodes::INVALID_TYPE,
          path: path,
          expected: "bool",
          received: Schema.type_name(value)
        )
      end
    end
  end
end
